File: Python-Src/node.py
from math import exp, pow
import random

class Node:

    def __init__(self, size, active_flag=1, name=''):
        self.name = name
        self.active_flag = active_flag
        self.basis = random.random()
        lst = []
        for index in range(0, size):
            lst.append(random.random())
        self.weights = lst
        self.front_pipes = []
        self.back_pipes = []

    def derive(self, learn_rate):
        t = self.__calculateT()
        outter_der = self.dev_active(t)
        param_der = 0.0
        for pipe in self.front_pipes:
            param_der += pipe.backward_result
        outter_der *= param_der
        for index in range(0, len(self.weights)):
            inner_der = self.back_pipes[index].forward_result
            self.weights[index] -= float(learn_rate * outter_der * inner_der)
            self.back_pipes[index].backward_result = outter_der * self.weights[index]
        self.basis -= float(learn_rate * outter_der)

    def calculate(self):
        t = self.__calculateT()
        for pipe in self.front_pipes:
            pipe.forward_result = self.active(t)

    def __calculateT(self):
        ans = float(self.basis)
        if len(self.back_pipes) != len(self.weights):
            print("This node is not working due to pipe and or weight configuration.")
        for index in range(0, len(self.weights)):
            ans += float(self.back_pipes[index].forward_result * self.weights[index])
        return ans

    def active(self, value):
        if self.active_flag == 1:
            return float(1.0 / (1.0 + exp(-value)))
        else:
            return 0.0

    def dev_active(self, value):
        if self.active_flag == 1:
            sig = self.active(value)
            return float(sig * (1.0 - sig))
        else:
            return 0.0

    def add_front_pipe(self, pipe):
        self.front_pipes.append(pipe)

    def add_back_pipe(self, pipe):
        self.back_pipes.append(pipe)

    def to_string(self):
        ans = "NODE " + self.name + "\n"
        ans += ("weights: " + str(self.weights)) + "\n"
        ans += ("basis: " + str(self.basis))
        return ans

File: Csharp-Src/Csharp-Src/Node.cs
using System;
using System.Collections.Generic;

namespace Csharp_Src
{
    public class Node
    {
        public string NodeName { get; set; }
        public int ActiveFlag { get; set; }
        public List<double> Weights { get; set; }
        public double Basis { get; set; }
        public List<Pipe> FrontPipes { get; set; }
        public List<Pipe> BackPipes { get; set; }

        public Node(int size, int activeFlag=1, string name="", bool isRandom = true)
        {
            Random rand = new Random();

            this.NodeName = name;
            this.ActiveFlag = activeFlag;
            this.Weights = new List<double>();

            this.Basis = isRandom ? rand.NextDouble() : 0.5;

            for (int i = 0; i < size; i++)
                this.Weights.Add(isRandom ? rand.NextDouble() : 0.5);

            this.FrontPipes = new List<Pipe>();
            this.BackPipes = new List<Pipe>();
        }

        private double CalculateT()
        {
            double ans = this.Basis;

            if (this.BackPipes.Count != this.Weights.Count)
                throw new Exception("This node is not working due to pipe and or weight configuration");
            
            for (int index = 0; index < this.Weights.Count; index++)
                ans += this.BackPipes[index].ForwardResult * this.Weights[index];
            
            return ans;
        }

        private double Active(double paramValue)
        {
            switch (this.ActiveFlag)
            {
                case 1:
                    return (double)(1.0 / (1.0 + Math.Exp(-paramValue)));
                default:
                    return 0.0;
            }
        }

        private double DeriveActive(double paramValue)
        {
            switch (this.ActiveFlag)
            {
                case 1:
                    double activeResult = this.Active(paramValue);
                    return (activeResult * (1.0 - activeResult));
                default:
                    return 0.0;
            }
        }

        public void Derive(double learningRate)
        {
            double tValue = this.CalculateT();
            double outterDer = this.DeriveActive(tValue);
            double paramDer = 0.0;

            foreach (var pipe in this.FrontPipes)
                paramDer += pipe.BackwardResult;

            outterDer *= paramDer;

            for (int index = 0; index < this.Weights.Count; index++)
            {
                double innerDer = this.BackPipes[index].ForwardResult;
                this.Weights[index] -= (double)(learningRate * outterDer * innerDer);
                this.BackPipes[index].BackwardResult = outterDer * this.Weights[index];
            }

            this.Basis -= learningRate * outterDer;
        }

        public void Calculate()
        {
            double tValue = this.CalculateT();
            
            foreach (var pip in this.FrontPipes)
                pip.ForwardResult = this.Active(tValue);
        }


        public override string ToString()
        {
            string ans = $"Node : {this.NodeName} \n";
            ans += $"weights: {this.Weights} \n";
            ans += $"basis: {this.Basis}";

            return ans;
        }
    }
}
